saleor/core/telemetry/metric.py

```diff
@@ -0,0 +1,205 @@
+        - core: For internal system operations and core functionality
+        - service: For business logic and service-level operations
+
+    Note:
+        The Meter internally uses the global OpenTelemetry MeterProvider, which should
+        be initialized following OpenTelemetry's standard process, for example using
+        the `opentelemetry-instrument` tool.
+
+    """
+
+    def __init__(self, instrumentation_version: str):
+        self._core_tracer = get_meter(CORE_SCOPE, instrumentation_version)
+        self._service_tracer = get_meter(SERVICE_SCOPE, instrumentation_version)
+        self._instruments: dict[str, tuple[Unit, Synchronous]] = {}
+        self._lock = Lock()
+
+    @staticmethod
+    def _create_instrument(
+        otel_meter: OtelMeter,
+        name: str,
+        type: MetricType,
+        unit: Unit,
+        description: str,
+    ) -> Synchronous:
+        kwargs = {"unit": unit.value, "description": description}
+        if type == MetricType.COUNTER:
+            return otel_meter.create_counter(name, **kwargs)
+        if type == MetricType.UP_DOWN_COUNTER:
+            return otel_meter.create_up_down_counter(name, **kwargs)
+        if type == MetricType.HISTOGRAM:
+            return otel_meter.create_histogram(name, **kwargs)
+        raise AttributeError(f"Unsupported instrument type: {type}")
+
+    def create_metric(
+        self,
+        name: str,
+        *,
+        type: MetricType,
+        unit: Unit,
+        service_scope: bool = False,
+        description: str = "",
+    ) -> None:
+        """Create a new metric with specified parameters.
+
+        Args:
+            name: The name of the metric
+            type: Type of the metric
+            unit: Unit of the metric
+            service_scope: If True, use service scope instead of default core
+            description: Optional description of the metric
+
+        """
+        meter = self._service_tracer if service_scope else self._core_tracer
+        with self._lock:
+            if name in self._instruments:
+                raise DuplicateMetricError(name)
+            instrument = self._create_instrument(meter, name, type, unit, description)
+            self._instruments[name] = unit, instrument
+
+    def record(
+        self,
+        metric_name: str,
+        amount: Amount,
+        *,
+        unit: Unit | None = None,
+        attributes: Attributes = None,
+    ) -> None:
+        """Record a measurement for the specified metric.
+
+        Args:
+            metric_name: Name of the metric to record a measurement
+            amount: Value to record
+            unit: Optional unit of the measurement (converted if different from metric unit)
+            attributes: Attributes to record with the measurement
+
+        """
+        attributes = enrich_with_global_attributes(attributes)
+        if metric_name not in self._instruments:
+            raise RuntimeError(f"Metric {metric_name} must be created first")
```

Comment:
We should also test this case.

Reply:
Yes, great point! We plan to add tests after confirming that the proposed API is acceptable, or in a separate PR to the feature branch